packages/text-area/src/TextArea.tsx

```diff
@@ -0,0 +1,278 @@
+import React, { useState, useMemo } from 'react';
+import { IdAllocator, Either, createDataProp } from '@leafygreen-ui/lib';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import { spacing, fontFamilies } from '@leafygreen-ui/tokens';
+
+const idAllocator = IdAllocator.create('textarea');
+const textAreaProp = createDataProp('area-selector');
+
+export const State = {
+  None: 'none',
+  Error: 'error',
+} as const;
+
+export type State = typeof State[keyof typeof State];
+
+const Mode = {
+  Light: 'light',
+  Dark: 'dark',
+} as const;
+
+type Mode = typeof Mode[keyof typeof Mode];
+
+const containerStyles = css`
+  display: flex;
+  flex-direction: column;
+`;
+
+const labelStyle = css`
+  font-size: 14px;
+  font-weight: bold;
+  line-height: 16px;
+  padding-bottom: 4px;
+`;
+
+const descriptionStyle = css`
+  font-size: 14px;
+  line-height: 16px;
+  font-weight: normal;
+  padding-bottom: 4px;
+  margin-top: 0px;
+  margin-bottom: 0px;
+`;
+
+const textAreaContainer = css`
+  margin-top: 4px;
+  position: relative;
+  display: flex;
+  align-items: center;
+  z-index: 0;
+`;
+
+const textAreaStyle = css`
+  width: 100%;
+  min-height: ${spacing[6]}px;
+  resize: none;
+  border-radius: 4px;
+  padding-left: 12px;
+  font-size: 14px;
+  font-weight: normal;
+  font-family: ${fontFamilies.default};
+  z-index: 1;
+
+  &:focus {
+    outline: none;
+  }
+
+  &:disabled {
+    cursor: not-allowed;
+  }
+`;
+
+const interactionRingStyles = css`
+  transition: all 150ms ease-in-out;
+  transform: scale(0.9, 0.8);
+  border-radius: 7px;
+  position: absolute;
+  top: -3px;
+  bottom: -3px;
+  left: -3px;
+  right: -3px;
+  pointer-events: none;
+  background-color: pink;
+
+  ${textAreaProp.selector}:hover ~ & {
+    transform: scale(1);
+  }
+
+  ${textAreaProp.selector}:focus ~ & {
+    transform: scale(1);
+  }
+`;
+
+const errorMessageStyle = css`
+  font-size: 14px;
+  height: 20px;
+  padding-top: 4px;
+  font-weight: normal;
+`;
+
+interface ColorSets {
+  labelColor: string;
+  descriptionColor: string;
+  textAreaBackgroundColor: string;
+  textAreaColor: string;
+  disabledBackgroundColor: string;
+  disabledColor: string;
+  interactionRing: string;
+  interactionRingFocus: string;
+  errorBorder: string;
+  errorMessage: string;
+}
+
+const colorSets: Record<Mode, ColorSets> = {
+  [Mode.Light]: {
+    labelColor: uiColors.gray.dark2,
+    descriptionColor: uiColors.gray.dark1,
+    textAreaColor: uiColors.gray.dark3,
+    textAreaBackgroundColor: uiColors.white,
+    disabledColor: uiColors.gray.base,
+    disabledBackgroundColor: uiColors.gray.light2,
+    interactionRing: uiColors.gray.light2,
+    interactionRingFocus: '#9dd0e7',
+    errorBorder: uiColors.red.base,
+    errorMessage: uiColors.red.base,
+  },
+  [Mode.Dark]: {
```

Comment:
I think I prefer the approach to modes and variants used in [Tabs](https://github.com/mongodb/leafygreen-ui/blob/master/packages/tabs/src/Tabs.tsx#L19) and [Toast](https://github.com/mongodb/leafygreen-ui/blob/master/packages/toast/src/Toast.tsx#L74) to this approach. I find this much harder to read, and I'd like to see us to consolidate our approach to style variants like this over time.

An added benefit to those approaches is not having to inline those stateful styles like you are now (something that I believe you've expressed disliking in the past).

Reply:
An object of objects you mean?

Reply:
An object of `css''` declarations so that you have the full declaration rather than just a color that's consumed elswehere for each state. Tabs is a good example.

Reply:
Happy to update

Reply:
```
interface ColorSets {
  labelColor: string;
  descriptionColor: string;
  textArea: {
    textAreaBackgroundColor: string;
    textAreaColor: string;
  };
  disabled: {
    backgroundColor: string;
    color: string;
  };
  interactionRing: {
    backgroundColor: string;
    focusColor: string;
  };
  error: {
    border: string;
    message: string;
  };
}```

Reply:
is that what you had in mind?

Reply:
More like:
```js
const modes =  {
  [Modes.Dark]: {
    label: css``,
    description: css``,
    textArea: {
      static: css``,
      disabled: css``,
    },
    ...etc
  }
}
```